tests/test_scheduler.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+from functools import partial
+from typing import List
+
+import pytest
+import torch
+from composer.core import State, Time, TimeUnit
+from composer.devices import DeviceCPU, DeviceGPU
+from composer.models import ComposerClassifier
+from composer.optim.scheduler import ComposerScheduler
+
+from llmfoundry.optim.scheduler import InverseSquareRootWithWarmupScheduler
+
+MAX_DURATION = '100ba'
+STEPS_PER_EPOCH = 100
+
+
+class SimpleModel(ComposerClassifier):
+    """Small classification model.
+
+    Args:
+        num_features (int): number of input features (default: 1)
+        num_classes (int): number of classes (default: 2)
+    """
+
+    def __init__(
+        self,
+        num_features: int = 1,
+        num_classes: int = 2,
+        num_hidden: int = 8,
+        device: str = 'cpu',
+        bias: bool = True,
+    ) -> None:
+
+        self.num_features = num_features
+        self.num_classes = num_classes
+
+        fc1 = torch.nn.Linear(num_features,
+                              num_hidden,
+                              device=device,
+                              bias=bias)
+        fc2 = torch.nn.Linear(num_hidden, num_classes, device=device, bias=bias)
+
+        net = torch.nn.Sequential(
+            torch.nn.AdaptiveAvgPool2d(1),
+            torch.nn.Flatten(),
+            fc1,
+            torch.nn.ReLU(),
+            fc2,
+            torch.nn.Softmax(dim=-1),
+        )
+        net.param_init_fn = self.param_init_fn
+        super().__init__(module=net, num_classes=num_classes)
+
+        # Important: It is crucial that the FC layers are bound to `self`
+        # for the optimizer surgery tests.
+        # These tests attempt to perform surgery on `fc1` layer, and we want
+        # to make sure that post-surgery, self.fc1 refers to the same parameters
+        # as self.net[1]
+        self.fc1 = fc1
+        self.fc2 = fc2
+
+    def param_init_fn(self, module: torch.nn.Module):
+        init_fn = partial(torch.nn.init.normal_, mean=0.0, std=0.1)
+
+        if isinstance(module, torch.nn.Linear):
+            init_fn(module.weight)
+            if module.bias is not None:
+                torch.nn.init.zeros_(module.bias)
```

Comment:
Can this be simplified?

Reply:
@dakinggg may have some comments on best practices for tests in this repo

Reply:
this model isn't actually used beyond passing it to the state constructor afaict, so i think you can just pass like `torch.nn.Linear(5, 5)` as the model arg and avoid this whole large class.

Reply:
Changed to torch.nn.Linear.